### ISSUE TITLE
Resizable BAR / SAM tanking CPU performance on VKD3D

### ISSUE THREAD

**Issue by Faugus** [HansKristian-Work/vkd3d-proton#1406]
RX 6800 XT / 5600X

I've been doing some tests and I got the conclusion that Resizable BAR / SAM hurts the CPU perfomance when running games on VKD3D.

This problem doesn't happen on DXVK or Vulkan Native.
Also this issue doesn't happen on Windows.

VKD3D games I tested: World of Warcraft, Uncharted 4, Spider-Man, Spider-Man Miles Morales, Resident Evil 2 Remake, Resident Evil Village...

I posted on Mesa's GitLab some tests and comparissons:
https://gitlab.freedesktop.org/mesa/mesa/-/issues/7993#note_1702565

**Comment by Mershl** [comment]
I can confirm the findings in Uncharted 4 on an Ryzen 3700x + RX7900XTX.

While SAM is enabled the game is weirdly CPU bound, weirdly because all cores show 40% usage with no core exceeding that. Disabling SAM (in the BIOS) leads to a much more healthy per core usage of 80% and a framerate uplift of +56% on my system.

Tested on an ASUS B450F, 32GB RAM, vkd3d 2.8. No OC or similar.

**Comment by Mershl** [comment]
Copied my results from mesa Gitlab:

| 3440x1440 Ultra | Uncharted 4 |
| ------ | ------ |
| SAM ON + `VKD3D_CONFIG=no_upload_hvv` | 78 FPS |
| SAM ON | 50 FPS |
| SAM OFF | 78 FPS |

Turning on/off Above 4G decoding and Resizable BAR individually did not show a difference so I merged them into "SAM ON". Reported BAR is `Detected VRAM RAM=24560M, BAR=32768M` for SAM ON, `Detected VRAM RAM=24560M, BAR=256M` for SAM OFF.

`no_upload_hvv` could become a default config for Uncharted 4 until the issue is resolved.

**Comment by Faugus** [comment]
Just tried `VKD3D_CONFIG=no_upload_hvv` with Uncharted and it fixed the issue!

I got a small FPS hit, but now SAM is working properly.

| Uncharted | Menu | Ingame |
| ------ | ------ | ------ |
| SAM ON + `VKD3D_CONFIG=no_upload_hvv` | 141 FPS | 138 FPS |
| SAM ON      | 146 FPS       | 88 FPS |
| SAM OFF | 132 FPS | 120 FPS |

**Comment by Mershl** [comment]
> but now SAM is working properly.

From my understanding using `no_upload_hvv` you're keeping vkd3d from using any BAR memory. So it's not actually "working", but you're avoiding this title-specific performance hit.

I'm really surprised you're seeing a benefit comparing `ON + no_upload_hvv` vs. `OFF`. I was not able to reproduce that.

**Comment by Faugus** [comment]
Do you have `Resizable BAR` option ON in BIOS?

If you do, try to disable it.

Leave only `Above 4G decoding` ON.

**Comment by Mershl** [comment]
Did a quick re-test. Tested on an ASUS B450F with this months latest BIOS (v5003).

| 3440x1440 Ultra | Uncharted 4 |
| ------ | ------ |
| `Above 4G decoding` + `Resizeable BAR` + `VKD3D_CONFIG=no_upload_hvv` | 78 FPS |
| `Above 4G decoding` + `Resizeable BAR` | 50 FPS |
| `Above 4G decoding` + `VKD3D_CONFIG=no_upload_hvv` |  78 FPS |
| `Above 4G decoding` | 51 FPS |
| everything OFF | 78 FPS |

**Comment by Faugus** [comment]
Maybe you're not seeing difference because you're running at 4K.

I'm running at 1440p.

**Comment by aufkrawall** [comment]
I too am seeing this in Uncharted 4, fps in main menu explode from 90fps CPU bound to 160fps GPU bound. Doesn't happen on Windows with native D3D12, or at least it's fast enough to be always GPU bound there.

**Comment by Billli11** [comment]
Dragon's Dogma 2 also suffer from this.

with AMD CPU 2700x
~40fps without `VKD3D_CONFIG=no_upload_hvv`  
~60fps with `VKD3D_CONFIG=no_upload_hvv`

**Comment by shelterx** [comment]
Nvidia RTX 4070- Control suffer from this too.
Having ReBar enabled limits the GPU usage to ~65% in Control, 
Disabling ReBar makes the GPU use all power (99%) and adds +15FPS in Control.

What's weird is that VKD3D_CONFIG=no_upload_hvv doesn't seem to do anything, I have to disable Rebar in Bios to actually see a difference.

**Comment by Mershl** [comment]
In Horizon Forbidden West it's the opposite.

3440x1440, Very High Preset, 5800x3D+7900XTX
`VKD3D_CONFIG not set`: 105FPS
`VKD3D_CONFIG=no_upload_hvv`: 94FPS

If no general solution can be found it might make sense to create game-specific `no_upload_hvv` defaults in `device.c` where it helps performance in most cases.

**Comment by shelterx** [comment]
Yes, I  agree. However it might be different results depending on Vendor abf/or driver.
The odd bird here is Control. I hope they figure that one out.

**Comment by Faugus** [comment]
One thing I noticed when having `VKD3D_CONFIG=no_upload_hvv` set is that a gameplay recording becomes jittery/stuttery.

**Comment by Cardiacman13** [comment]
I can confirm this bug, without `VKD3D_CONFIG=no_upload_hvv` DD2 and some other games are unplayable.

**Comment by Billli11** [comment]
look like Warhammer 40,000: Space Marine 2 also affected by this.

Have ~20% fps improve with `no_upload_hvv` when cpu bound.

**Comment by mhmarf** [comment]
I just tested Warhammer 40,000: Space Marine 2 with `no_upload_hvv` and it massively REDUCED my FPS (from 100 to 88).
I am on 6900 xt ...

**Comment by Billli11** [comment]
@mhmarf  Is the game CPU bound?  
`no_upload_hvv` can help with CPU performance but wrost gpu performance.

**Comment by mhmarf** [comment]
Indeed you are correct. I tried hard to make it CPU bound but no matter where I looked, GPU was always 98%... therefore I lowered resolution to 800x600 and yes, indeed it reduced my FPS...

800x600:
no_upload_hvv -> 117 FPS
normal (without no_upload_hvv) -> 100 FPS

2560x1440:
no_upload_hvv -> 88 FPS
normal (without no_upload_hvv)  -> 100 FPS

**Comment by Arcitec** [comment]
Windows has the same exact problem. NVIDIA drivers contain whitelists about which games are **allowed** to use ReSIZABLE BAR. It's off by default for every game because it can hurt performance.
